File: src/common/Helpers/TryCatchHelpers.cs
using System;
using System.Collections.Generic;

public class TryCatchHelpers
{
    public static Action? NoThrowWrap(Action? action)
    {
        return action == null ? null : () =>
        {
            try { action(); }
            catch { }
        };
    }

    public static Action<T>? NoThrowWrap<T>(Action<T>? action)
    {
        return action == null ? null : (x =>
        {
            try { action(x); }
            catch { }
        });
    }

    public static Action<T1, T2>? NoThrowWrap<T1, T2>(Action<T1, T2>? action)
    {
        return action == null ? null : (x, y) =>
        {
            try { action(x, y); }
            catch { }
        };
    }

    public static Action<T1, T2, T3>? NoThrowWrap<T1, T2, T3>(Action<T1, T2, T3>? action)
    {
        return action == null ? null : (x, y, z) =>
        {
            try { action(x, y, z); }
            catch { }
        };
    }

    public static Action<T1, T2, T3, T4>? NoThrowWrap<T1, T2, T3, T4>(Action<T1, T2, T3, T4>? action)
    {
        return action == null ? null : (x, y, z, a) =>
        {
            try { action(x, y, z, a); }
            catch { }
        };
    }

    public static Func<(T? value, Exception? ex)> NoThrowWrapException<T>(Func<T> function, T defaultResult)
    {
        return () =>
        {
            try
            {
                return (function(), null);
            }
            catch (Exception ex)
            {
                return (defaultResult, ex);
            }
        };
    }

    public static Func<T1, (T2? value, Exception? ex)> NoThrowWrapException<T1, T2>(Func<T1, T2> function, T2 defaultResult)
    {
        return (x) =>
        {
            try
            {
                return (function(x), null);
            }
            catch (Exception ex)
            {
                return (defaultResult, ex);
            }
        };
    }

    public static Func<T1, T2, (T3? value, Exception? ex)> NoThrowWrapException<T1, T2, T3>(Func<T1, T2, T3> function, T3 defaultResult)
    {
        return (x, y) =>
        {
            try
            {
                return (function(x, y), null);
            }
            catch (Exception ex)
            {
                return (defaultResult, ex);
            }
        };
    }

    public static Func<T1, T2, T3, (T4? value, Exception? ex)> NoThrowWrapException<T1, T2, T3, T4>(Func<T1, T2, T3, T4> function, T4 defaultResult)
    {
        return (x, y, z) =>
        {
            try
            {
                return (function(x, y, z), null);
            }
            catch (Exception ex)
            {
                return (defaultResult, ex);
            }
        };
    }

    public static Func<T1, T2, T3, T4, (T5? value, Exception? ex)> NoThrowWrapException<T1, T2, T3, T4, T5>(Func<T1, T2, T3, T4, T5> function, T5 defaultResult)
    {
        return (x, y, z, a) =>
        {
            try
            {
                return (function(x, y, z, a), null);
            }
            catch (Exception ex)
            {
                return (defaultResult, ex);
            }
        };
    }

    public static T TryCatchNoThrow<T>(Func<T> function, T defaultResult, out Exception? functionThrewException)
    {
        functionThrewException = null;
        try
        {
            return function();
        }
        catch (Exception ex)
        {
            functionThrewException = ex;
        }
        return defaultResult;
    }
}

File: src/commonjs/Helpers/TryCatchHelpers.ts
export class TryCatchHelpers {
    public static NoThrowWrap(action: (() => void) | null): (() => void) | null {
        return action == null ? null : () => {
            try { action(); }
            catch { }
        };
    }

    public static NoThrowWrapT<T>(action: ((x: T) => void) | null): ((x: T) => void) | null {
        return action == null ? null : (x => {
            try { action(x); }
            catch { }
        });
    }

    public static NoThrowWrapTT<T1, T2>(action: ((x: T1, y: T2) => void) | null): ((x: T1, y: T2) => void) | null {
        return action == null ? null : (x, y) => {
            try { action(x, y); }
            catch { }
        };
    }

    public static NoThrowWrapTTT<T1, T2, T3>(action: ((x: T1, y: T2, z: T3) => void) | null): ((x: T1, y: T2, z: T3) => void) | null {
        return action == null ? null : (x, y, z) => {
            try { action(x, y, z); }
            catch { }
        };
    }

    public static NoThrowWrapTTTT<T1, T2, T3, T4>(action: ((x: T1, y: T2, z: T3, a: T4) => void) | null): ((x: T1, y: T2, z: T3, a: T4) => void) | null {
        return action == null ? null : (x, y, z, a) => {
            try { action(x, y, z, a); }
            catch { }
        };
    }

    public static NoThrowWrapException<T>(func: () => T, defaultResult: T): () => { value: T; ex: Error | null } {
        return () => {
            try {
                return { value: func(), ex: null };
            }
            catch (ex) {
                return { value: defaultResult, ex: ex as Error };
            }
        };
    }

    public static NoThrowWrapExceptionT<T1, T2>(func: (x: T1) => T2, defaultResult: T2): (x: T1) => { value: T2; ex: Error | null } {
        return (x) => {
            try {
                return { value: func(x), ex: null };
            }
            catch (ex) {
                return { value: defaultResult, ex: ex as Error };
            }
        };
    }

    public static NoThrowWrapExceptionTT<T1, T2, T3>(func: (x: T1, y: T2) => T3, defaultResult: T3): (x: T1, y: T2) => { value: T3; ex: Error | null } {
        return (x, y) => {
            try {
                return { value: func(x, y), ex: null };
            }
            catch (ex) {
                return { value: defaultResult, ex: ex as Error };
            }
        };
    }

    public static NoThrowWrapExceptionTTT<T1, T2, T3, T4>(func: (x: T1, y: T2, z: T3) => T4, defaultResult: T4): (x: T1, y: T2, z: T3) => { value: T4; ex: Error | null } {
        return (x, y, z) => {
            try {
                return { value: func(x, y, z), ex: null };
            }
            catch (ex) {
                return { value: defaultResult, ex: ex as Error };
            }
        };
    }

    public static NoThrowWrapExceptionTTTT<T1, T2, T3, T4, T5>(func: (x: T1, y: T2, z: T3, a: T4) => T5, defaultResult: T5): (x: T1, y: T2, z: T3, a: T4) => { value: T5; ex: Error | null } {
        return (x, y, z, a) => {
            try {
                return { value: func(x, y, z, a), ex: null };
            }
            catch (ex) {
                return { value: defaultResult, ex: ex as Error };
            }
        };
    }

    public static TryCatchNoThrow<T>(func: () => T, defaultResult: T): { result: T; exception: Error | null } {
        try {
            return { result: func(), exception: null };
        }
        catch (ex) {
            return { result: defaultResult, exception: ex as Error };
        }
    }
}